Feat/308 send duration of mem cache to metrics

taddes opened the pull request:
This task added the `duration` variable - `Instant::now() - start` during memory cleanup phase so it can be ingested and viewed in our metrics. The `duration` variable is passed into the `.time_macros()` method under an instance of the `metrics_client`. 

The `time_macros()` function takes a a key and value pair and is capable of processing the duration struct, not requiring conversion using the rust time crate. The `.ok()` call at the end simply handles the conversion from `Result<T, E>` to the `Option<T>` to ensure the variable does not raise a compilation error of an unused variable. There was no tag addition required for this.

Also added documentation entries in the `data.md` file so the new key, as well as a missing key were added.

pjenvey commented on merino-cache/src/memory.rs:
```diff
@@ -191,6 +191,9 @@ impl Suggester {
             .gauge("cache.memory.storage-len", self.items.len_storage() as u64)
             .ok();
+        self.metrics_client
+            .time_micros("cache.memory.duration-ms", duration)
```
You want the plain `time` method for millisecond reporting as `time_micros` reports microseconds instead.

FYI all the timers in merino currently are measuring Provider response times. They're all basically in-memory providers at this point which return responses very quickly (sub millisecond), hence the `time_micros` usage everywhere.

However I think you're right to prefer millis here -- after populating the memory provider locally via the load tester I see the removal task taking around 2-12 ms to complete.

pjenvey replied:
Adjusting the `time` call hopefully fixes the failing test. Additionally, we might as well add a new assertion for the new metric in that same test `cache_maintainer_removes_expired_entries` -- probably a one liner

taddes replied:
Gotcha. Thanks for discussing. Would love to go over the testing you did locally to get that metric value. It appears I have to implement the StatsdClient differently as I cannot access the `time` variable, solution looks to be referenced in `client.rs`. I'm poking around right now, but don't want to change a major implementation detail without you feedback first. Appears to be boxed already, but somewhat differently than in the client.rs.

https://github.com/mozilla-services/merino/blob/cfdd68dc6f6461e583241bc0f0097947199d986f/merino-cache/src/memory.rs#L103 

`/// Trait that encompasses all other traits for sending metrics.
///
/// If you wish to use `StatsdClient` with a generic type or place a
/// `StatsdClient` instance behind a pointer (such as a `Box`) this will allow
/// you to reference all the implemented methods for recording metrics, while
/// using a single trait. An example of this is shown below.
///
/// ```
/// use std::time::Duration;
/// use cadence::{MetricClient, StatsdClient, NopMetricSink};
///
/// let client: Box<dyn MetricClient> = Box::new(StatsdClient::from_sink(
///     "prefix", NopMetricSink));
/// ...
/// client.time("some.timer", 42).unwrap();
/// client.time("some.timer", Duration::from_millis(42)).unwrap();

/// ````

jrconlin replied:
I think this might be a bit easier than that.

The error you might be getting is something like:
```rust
error[E0599]: no method named `time_with_tags` found for struct `StatsdClient` in the current scope
   --> merino-cache/src/memory.rs:297:22
    |
297 |                     .time_with_tags("cache.memory.duration", now.elapsed())
    |                      ^^^^^^^^^^^^^^ method not found in `StatsdClient`
    |
   ::: /home/jrconlin/.cargo/registry/src/github.com-1ecc6299db9ec823/cadence-0.26.0/src/client.rs:274:8
    |
274 |     fn time_with_tags<'a>(&'a self, key: &'a str, time: T) -> MetricBuilder<'_, '_, Timer>;
    |        -------------- the method is available for `StatsdClient` here
    |
    = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
    |
8   | use cadence::Timed;
    |

For more information about this error, try `rustc --explain E0599`.
```

Adding the `use` brings the `Timed` trait in context. This helps the compiler find the right methods it needs so you can do 

```rust
                self.metrics_client
                    .time_with_tags("cache.memory.duration", now.elapsed())
                    .with_tag("cache-status", response.cache_status.to_string().as_str())
                    .send();
```
that should let you compile without significant changes. 

(Mind you, locally I'm getting odd internal actix build errors which I think might be related to a version conflict triggering inside of the common-rs library. Hopefully, you're not getting that.)

taddes replied:
Thanks @jrconlin , I was getting some build errors that were different, but mostly due to the M1 architecture on my laptop which related to something not building locally, but that was solved when committed and pushed. But yes, the addition of the use statement allowed the use of `time` which can then be ingested by StatsdClient, without having to add the extra tag.

pjenvey commented on docs/data.md:
```diff
@@ -125,9 +125,16 @@ log inspection interfaces.
 - `cache.memory.storage-len` - A gauge representing the number of entries in the
   second level of hashing in the in-memory deduped hashmap.
 
+- `cache.memory.remove-expired` - A record of expired entries for tracing framework 
```
Thanks for catching this -- however it's a log message (and of an INFO level) so it belongs in the Logs section above.

taddes replied:
Thank you! I moved it to the log section, appreciate the semantic clarity, totally missed that

pjenvey commented:
Also you're hitting a brand new Rust CVE in the checks failure (from cargo audit). I don't think this CVE actually affects Merino. 

We could fix this in a separate PR but it's easy enough for you to fix by upgrading the regex crate via `cargo update -p regex`

pjenvey commented on merino-cache/src/memory.rs:
```diff
@@ -379,15 +382,24 @@ mod tests {
-        assert_eq!(rx.len(), 2);
+        assert_eq!(rx.len(), 3);
         let collected_data: Vec<String> = rx
             .iter()
```
You've hit a footgun w/ this channel `Receiver` from `SpyMetrics` -- it's a ["blocking" iterator](https://docs.rs/crossbeam-channel/latest/crossbeam_channel/struct.Receiver.html#method.iter), where "_Each call to [next](https://doc.rust-lang.org/nightly/core/iter/traits/iterator/trait.Iterator.html#tymethod.next) blocks waiting for the next message and then returns it._".

The `take(2)` call issued to the iterator previously avoided its blocking (and we proved there was only 2 messages to take by asserting so above). You've removed that call, so the iteration continues instead of stopping -- `collect` gets the 3rd message but issues another `next` call. Instead of signaling the Iterator's exhausted (returning `None` to `next`) it blocks, waiting for the other side of the channel to send a new value which never happens. Hence the test timing out in CI.

Issuing `take(3)` fixes this but there's a better fix I've been meaning to apply everywhere:

```suggestion
            .try_iter()
```

This uses a [non-blocking `Iterator`](https://docs.rs/crossbeam-channel/latest/crossbeam_channel/struct.Receiver.html#method.try_iter) instead so we don't have to limit the iteration via `take`. 

Can you switch to `try_iter` here and additionally fix `merino-suggest-providers/src/providers/keyword_filter.rs` (its `test_provider_all_filtered` also relies on `take()`, let's get rid of it)

taddes replied:
Fascinating @pjenvey ! @ncloudioj was incredibly generous with his time today to give me an excellent overview of iterators, adapters and their properties so I learned a ton and this adds some interesting extra nuances of detail.  That very clearly explains the blocking/non-blocking nature of the two methods which I will be sure to make note of!

The removal of `take()` originally was intended to remove a redundant method, given we had just above checked that there were 3 dummy elements in that vector. However, upon your explanation, I better understand why it was there. I will make those adjustments now.

taddes replied:
Indeed, I have no idea how it passed on my machine @ncloudioj .Maybe the M1 architecture is just so good, it fixes compiler errors on the fly 😆 Either way, thank you @pjenvey  for catching that. Glad it resolves something you just found out about!

taddes commented:
Just going to summarize the changes here for clarity (and for learning purposes):
1.  Added `Timed` in use statement for `cadence` to access `time()` method off of `metrics_client` so the duration in miliseconds can be ingested with ease. No need for microsecond usage or any additional tags here.
`        self.metrics_client
            .time("cache.memory.duration", duration)
            .ok();`
2. In the testing function `cache_maintainer_removes_expired_entries`, several modifications were made:
- the assertion statement to enumerate the length of the `Reciever<Vec<u8>>` updated to 3 to include new metric 
- removed the `take()` method in iterator adapter expression, but as a result have to update to `try_iter()` as @pjenvey  recommended, due to blocking issue if `take()` is removed. Originally, `take()` removed, as it seemed it was related to ensuring 3 values are utilized for testing, which seemed redundant due to existing `assert_eq!(rx.len(), 3);`
- modified the handling of how the metrics are verified by calling `next()` on the `collected_data` iterator. Finally, to avoid a regular expression or issue of possible metrics values differing, simply calling `starts_with` and `ends_with` on the string value of vec index results in a match that doesn't depend on the many possibilities of time values. 
3. In `merino-suggest-providers/src/providers/keyword_filter.rs`, updated `take` to implement same `try_iter()`logic

ncloudioj reviewed (approved):
Looks good, thanks! @taddes Wooh, first patch in Rust! \o/ 🎊

ncloudioj commented on merino-cache/src/memory.rs:
```diff
@@ -379,15 +382,24 @@ mod tests {
+            .try_iter()
             .map(|x| String::from_utf8(x).unwrap())
-            .collect();
+            .collect::<Vec<String>>();
```
Nit: `::<Vec<string>>` is not necessary here as it's already specified above by `collected_data`.

taddes replied:
Yes, I recall we discussed it @ncloudioj , I had it there to be extra explicit, but I understand it's definitely not required and redundant.

pjenvey commented on docs/data.md:
```diff
@@ -125,6 +129,9 @@ log inspection interfaces.
 - `cache.memory.pointers-len` - A gauge representing the number of entries in
   the first level of hashing in the in-memory deduped hashmap.
 
+- `cache.memory.duration-ms` - A duration in milliseconds representing the time 
```
```suggestion
- `cache.memory.duration` - A duration in milliseconds representing the time 
```

taddes replied:
Updated
